### PR TITLE
Prime all user meta caches

### PR DESCRIPTION
**Pull request by spacedmonkey**
Trac ticket: https://core.trac.wordpress.org/ticket/55594

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Comment by TimothyBJacobs** [comment]
I'm not entirely following the patch here. Can you expand on the what the actual changes are here and their purpose? At first glance, it looks to be collapsing `all` and `all_with_meta` to be equivalent. Is that correct?

Related: https://core.trac.wordpress.org/ticket/22057

**Comment by spacedmonkey** [comment]
@TimothyBJacobs In WordPress 4.9, the WP_User class was changed so needs user meta. See 
https://github.com/WordPress/wordpress-develop/blob/4e830858061dbe37cda462e9a861bc7401fc6f6b/src/wp-includes/class-wp-user.php#L890-L906

So in the user REST API controller and is all other places that load a list of users, user meta is not primed correctly.

**Comment by TimothyBJacobs** [comment]
So should we update the docs for `WP_User_Query` to indicate that these function identically?

**Comment by peterwilsoncc** [comment]
Thanks for explaining this, I was a little confused too. I can see why this is needed for both now with the changes to `WP_User`. 

The `all_with_meta` option was effectively soft deprecated in 4.9 once `all` started populating the meta data. It could be simpler to document & throw a notice for the soft deprecation and convert it to `all`.

File: src/wp-includes/class-wp-user-query.php [hunk 1]
```diff
@@ -835,26 +833,27 @@ public function query() {
 		if ( ! $this->results ) {
 			return;
 		}
+
```
Comment (spacedmonkey) [inline src/wp-includes/class-wp-user-query.php]:
```suggestion
```

**Comment by spacedmonkey** [comment]
> The all_with_meta option was effectively soft deprecated in 4.9 once all started populating the meta data. It could be simpler to document & throw a notice for the soft deprecation and convert it to all.

I have added a comment, to deprecate. I don't think we should hard deprecate, this param is used in a number of places in core and I guess in lots plugins as well.

File: src/wp-includes/class-wp-user-query.php [hunk 2]
```diff
@@ -842,19 +840,19 @@ public function query() {
...
 			cache_users( $this->results );
 
 			$r = array();
 			foreach ( $this->results as $userid ) {
-				$r[ $userid ] = new WP_User( $userid, '', $qv['blog_id'] );
+				if ( 'all_with_meta' === $qv['fields'] ) {
+					$r[ $userid ] = new WP_User( $userid, '', $qv['blog_id'] );
+				} else {
+					$r[] = new WP_User( $userid, '', $qv['blog_id'] );
+				}
```
Comment (peterwilsoncc) [inline src/wp-includes/class-wp-user-query.php:852]: Is this the code that prevents you from normalising `all_with_fields` to `all` at the start of the function?

Comment (spacedmonkey) [reply src/wp-includes/class-wp-user-query.php:852]: I am not sure of the question. This code ensure that `all_with_meta` returns an array of with key of user id. Whereas `all` return an array of WP_User objects with a numeric key.

File: src/wp-includes/class-wp-user-query.php [hunk 3]
```diff
@@ -310,9 +310,7 @@ public function prepare_query( $query = array() ) {
 				$this->query_fields[] = "$wpdb->users.$field";
 			}
 			$this->query_fields = implode( ',', $this->query_fields );
-		} elseif ( 'all' === $qv['fields'] ) {
-			$this->query_fields = "$wpdb->users.*";
-		} elseif ( ! in_array( $qv['fields'], $allowed_fields, true ) ) {
+		} elseif ( 'all_with_meta' === $qv['fields'] || 'all' === $qv['fields'] || ! in_array( $qv['fields'], $allowed_fields, true ) ) {
 			$this->query_fields = "$wpdb->users.ID";
```
Comment (spacedmonkey) [inline src/wp-includes/class-wp-user-query.php:314]: We can now just get the ID of the user, as `cache_users` is called before, priming the user caches.

File: src/wp-includes/class-wp-user-query.php (hunk 2 above)
Comment (peterwilsoncc) [reply src/wp-includes/class-wp-user-query.php:852]: > This code ensure that `all_with_meta` returns an array of with key of user id. Whereas `all` return an array of WP_User objects with a numeric key.

That answers my question, thanks.

**Review by mehulkaklotar — APPROVED** [review]
LGTM 🚀

**Review by peterwilsoncc — APPROVED** [review]
This looks good to me.

I've merged in `trunk` to make sure all the tests are still passing, the `users` group is doing so locally so I expect that the complete test suite should be dandy too.

**Comment by spacedmonkey** [comment]
Commited https://github.com/WordPress/wordpress-develop/commit/2db7debe8a2f0b2ea264d1765c549bbf5b44a052

**Pull request closed without merging**